File: testbed/test/cruise/statemachine/test/EntryAndExitActionTest.java
package cruise.statemachine.test;

import org.junit.*;

public class EntryAndExitActionTest
{
  
  @Test
  public void EntryCalledOnConstructorForDefault()
  {
    CourseB course = new CourseB();
    Assert.assertEquals("entry called",course.getLog());
  }
  
  @Test
  public void CallEntry()
  {
    CourseB course = new CourseB();
    course.anEvent();
    course.anEvent();
    Assert.assertEquals(CourseB.Status.Open,course.getStatus());
    Assert.assertEquals("entry called", course.getLog());
  }  
  
  @Test
  public void CallExit()
  {
    CourseB course = new CourseB();
    course.anEvent();
    Assert.assertEquals(CourseB.Status.Closed,course.getStatus());
    Assert.assertEquals("exit called", course.getLog());
    
    course.anEvent();
    Assert.assertEquals(CourseB.Status.Open, course.getStatus());
    Assert.assertEquals("entry called", course.getLog());
  }
  
  @Test
  public void CallMultipleEntryExit()
  {
    CourseS course = new CourseS();
    
    Assert.assertEquals("Enter Off 1", course.getLog(0));
    Assert.assertEquals("Enter Off 2", course.getLog(1));

    course.flip();
    Assert.assertEquals("Exit Off 1", course.getLog(2));
    Assert.assertEquals("Exit Off 2", course.getLog(3));
  }

  @Test
  public void ExitActionSelfTransition()
  {
    ExitActionSelfTransition sm = new ExitActionSelfTransition();
	Assert.assertEquals(ExitActionSelfTransition.Sm.created, sm.getSm());
	sm.init();
	Assert.assertEquals(ExitActionSelfTransition.Sm.created, sm.getSm());
	Assert.assertTrue(sm.getExitCodeCalled());
  }
}

File: testbed_python/test/statemachine/entry_and_exit_action_test.py
import unittest

# Test classes are imported dynamically
# Must be generated into ImportModules' namespace then imported
from ImportModules import importModules

importModules(
    ["CourseB", "CourseS", "ExitActionSelfTransition"], ["statemachine", "test"]
)
from ImportModules import *

import CourseB
import CourseS
import ExitActionSelfTransition


class EntryAndExitActionTest(unittest.TestCase):
    def EntryCalledOnConstructorForDefault(self):
        course = CourseB.CourseB()
        self.assertEquals("entry called", course.getLog())

    def CallEntry(self):
        course = CourseB.CourseB()
        course.anEvent()
        course.anEvent()
        self.assertEquals(CourseB.CourseB.Status.Open, course.getStatus())
        self.assertEquals("entry called", course.getLog())

    def CallExit(self):
        course = CourseB.CourseB()
        course.anEvent()
        self.assertEquals(CourseB.CourseB.Status.Closed, course.getStatus())
        self.assertEquals("exit called", course.getLog())

        course.anEvent()
        self.assertEquals(CourseB.CourseB.Status.Open, course.getStatus())
        self.assertEquals("entry called", course.getLog())

    def CallMultipleEntryExit(self):
        course = CourseS.CourseS()

        self.assertEquals("Enter Off 1", course.getLog(0))
        self.assertEquals("Enter Off 2", course.getLog(1))

        course.flip()
        self.assertEquals("Exit Off 1", course.getLog(2))
        self.assertEquals("Exit Off 2", course.getLog(3))

    def ExitActionSelfTransition(self):
        sm = ExitActionSelfTransition.ExitActionSelfTransition()
        self.assertEquals(
            ExitActionSelfTransition.ExitActionSelfTransition.Sm.created, sm.getSm()
        )
        sm.init()
        self.assertEquals(
            ExitActionSelfTransition.ExitActionSelfTransition.Sm.created, sm.getSm()
        )
        self.assertTrue(sm.getExitCodeCalled())
